### ISSUE TITLE
Please merge dart2js changes into stable channel 2.7.*

### ISSUE THREAD

**Issue by sigmundch** [dart-lang/sdk#40217]
**commit(s) to merge**: aa56cd39a3 1103d9fdb5 b050a59d3b a57a97703a 75fc15c7e1 b233061a90 4bc9c8b52c 529913fd39 a92a9f1ba2

**merge instructions**: 

```
git cherry-pick aa56cd39a3 1103d9fdb5 b050a59d3b a57a97703a 75fc15c7e1 b233061a90 4bc9c8b52c 529913fd39 a92a9f1ba2
```

Together these merged cleanly on my client and I was able to build the sdk and manually run dart2js.

**reason**: considerable compile-time performance improvements for external apps (details below). We iterated on these changes for a while with one our customers, they have tried dev releases and patches with these changes and and they saw a huge reduction in memory utilization on their apps (from 20G to 12G of RAM), and compilation time (from 35 minutes, down to 11 minutes). These changes were designed to be generally helpful to other apps as well.

/cc @dgrove @vsm @kevmoo @mit-mit @whesse @athomas 
/cc @rakudrama 

**Details**

We have been iterating for a long time on reducing the total memory used and total compilation time spent on some large apps. Some apps have more pathological cases than others (see #27883).

We made several changes to help with this, and some small refactoring in between that unfortunately we need to also cherry-pick in order to make the merge clean.

*a92a9f1ba2 (dart2js) initialize the receiver type to its static type during inference*:
  * important memory and compile-time savings: about 25-30% memory savings on small samples, possible more for the large apps, and 2x performance improvement.

*529913fd39 (dart2js) remove unnecessary uses of receiverMask*:
*4bc9c8b52c (dart2js) rename typeOfSend, delete unused typeOfGetter*:
  * both are small refactorings to make the change above easier

*b233061a90 Ensure native methods default to have all side-effects*:
  * correctness issue: addressing the performance issue exposes a correctness issue fixed here first.

*4eae39672a (dart2js) release some data-structures after resolution*:
  * a large memory savings as well, about 25% of memory on some samples.

*75fc15c7e1 [dart2js] Use cache in CommonMasks.isInterceptor*:
  * an important compilation performance improvement, cuts down about 500s to 10s on large apps.

*a57a97703a [dart2js] Make repeated 'usedAsVariable' tests faster*:
 * another compilation performance improvement, cuts down about 100s of compilation time on pathological apps.

*b050a59d3b (dart2js) merge inferrer-engine interface and implementation*:
*1103d9fdb5 (dart2js): seal-in members used only within the inferrer_engine class*:
*aa56cd39a3 (dart2js): rename 'assignments' to 'inputs' in global analysis*:
  * refactor needed to make later merge clean

**Comment by robbecker-wf** [comment]
FYI I verified that these cherry picks onto the 2.7.0 tag exhibit all of the dart2js performance gains for us.

**Comment by athomas** [comment]
Released with 2.7.1.
